File: python/src/nr_ai_agent/metrics/cost_attribution.py
"""Cost attribution via context variable propagation and per-request metadata."""

from contextvars import ContextVar
from typing import Any, Dict, Optional

_attribution_context: ContextVar[Optional[Dict[str, str]]] = ContextVar(
    "nr_attribution_context", default=None
)


def set_attribution_context(tags: Dict[str, str]) -> None:
    """Set attribution tags for the current context (thread/async task)."""
    _attribution_context.set(tags)


def get_attribution_context() -> Optional[Dict[str, str]]:
    """Get attribution tags for the current context."""
    return _attribution_context.get()


def clear_attribution_context() -> None:
    """Clear attribution tags for the current context."""
    _attribution_context.set(None)


def resolve_attribution(
    request_metadata: Optional[Dict[str, Any]],
    context_tags: Optional[Dict[str, str]],
    global_tags: Optional[Dict[str, str]],
) -> Dict[str, str]:
    """Merge attribution tags: per-request > context > global."""
    merged: Dict[str, str] = {}

    # Apply global defaults first
    if global_tags:
        for key, value in global_tags.items():
            if value is not None:
                merged[key] = value

    # Apply context tags (override globals)
    if context_tags:
        for key, value in context_tags.items():
            if value is not None:
                merged[key] = value

    # Extract per-request tags from metadata.nr.*
    if request_metadata and isinstance(request_metadata, dict):
        nr = request_metadata.get("nr")
        if nr and isinstance(nr, dict):
            for key, value in nr.items():
                if isinstance(value, str):
                    merged[key] = value

    return merged


def attribution_tags_to_custom_attributes(tags: Dict[str, str]) -> Dict[str, Any]:
    """Convert attribution tags to flat NR custom attribute dict."""
    attrs: Dict[str, Any] = {}

    if tags.get("feature"):
        attrs["ai.attribution.feature"] = tags["feature"]
    if tags.get("team"):
        attrs["ai.attribution.team"] = tags["team"]
    if tags.get("user"):
        attrs["ai.attribution.user"] = tags["user"]
    if tags.get("environment"):
        attrs["ai.attribution.environment"] = tags["environment"]

    standard_keys = {"feature", "team", "user", "environment"}
    for key, value in tags.items():
        if key not in standard_keys and value is not None:
            attrs[f"ai.custom.{key}"] = value

    return attrs


def strip_nr_metadata(metadata: Any) -> Any:
    """Remove nr.* metadata from a request metadata dict."""
    if not metadata or not isinstance(metadata, dict):
        return metadata
    if "nr" in metadata:
        return {k: v for k, v in metadata.items() if k != "nr"}
    return metadata

File: packages/nr-ai-agent/src/metrics/cost-attribution.ts
export interface AttributionTags {
  readonly feature?: string;
  readonly team?: string;
  readonly user?: string;
  readonly environment?: string;
  readonly [key: string]: string | undefined;
}

export function resolveAttribution(
  requestMetadata: Record<string, unknown> | null | undefined,
  contextTags: AttributionTags | undefined,
  globalTags: AttributionTags | undefined,
): AttributionTags {
  // Extract per-request tags from metadata.nr.*
  const requestTags: Record<string, string> = {};
  if (requestMetadata && typeof requestMetadata === 'object' && 'nr' in requestMetadata) {
    const nr = requestMetadata.nr;
    if (nr && typeof nr === 'object') {
      for (const [key, value] of Object.entries(nr)) {
        if (typeof value === 'string') {
          requestTags[key] = value;
        }
      }
    }
  }

  // Merge: per-request > context > global (specific overrides general)
  const merged: Record<string, string> = {};

  // Apply global defaults first
  if (globalTags) {
    for (const [key, value] of Object.entries(globalTags)) {
      if (value !== undefined) {
        merged[key] = value;
      }
    }
  }

  // Apply context tags (override globals)
  if (contextTags) {
    for (const [key, value] of Object.entries(contextTags)) {
      if (value !== undefined) {
        merged[key] = value;
      }
    }
  }

  // Apply per-request tags (override both)
  for (const [key, value] of Object.entries(requestTags)) {
    if (value !== undefined) {
      merged[key] = value;
    }
  }

  return merged as AttributionTags;
}

export function attributionTagsToCustomAttributes(tags: AttributionTags): Record<string, string | number> {
  const attrs: Record<string, string | number> = {};

  // Standard attribution tags
  if (tags.feature) {
    attrs['ai.attribution.feature'] = tags.feature;
  }
  if (tags.team) {
    attrs['ai.attribution.team'] = tags.team;
  }
  if (tags.user) {
    attrs['ai.attribution.user'] = tags.user;
  }
  if (tags.environment) {
    attrs['ai.attribution.environment'] = tags.environment;
  }

  // Custom tags with ai.custom.* prefix
  for (const [key, value] of Object.entries(tags)) {
    if (
      key !== 'feature' &&
      key !== 'team' &&
      key !== 'user' &&
      key !== 'environment' &&
      value !== undefined
    ) {
      attrs[`ai.custom.${key}`] = value;
    }
  }

  return attrs;
}

export function stripNrMetadata(metadata: unknown): unknown {
  if (!metadata || typeof metadata !== 'object') {
    return metadata;
  }

  const meta = metadata as Record<string, unknown>;

  // If metadata has nr field, create a copy without it
  if ('nr' in meta) {
    const { nr: _nr, ...rest } = meta;
    return rest;
  }

  return metadata;
}
